src/txrequest.cpp

```diff
@@ -0,0 +1,577 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txrequest.h>
+
+#include <crypto/siphash.h>
+#include <net.h>
+#include <primitives/transaction.h>
+#include <random.h>
+#include <uint256.h>
+#include <util/memory.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+
+#include <assert.h>
+
+namespace {
+
+/** The various states a (txhash,peer) pair can be in.
+ *
+ * Note that CANDIDATE is split up into 3 substates (DELAYED, BEST, READY), allowing more efficient implementation.
+ * Also note that the sorting order of ByTxHashView relies on the specific order of values in this enum.
+ *
+ * Expected behaviour is:
+ *   - When first announced by a peer, the state is CANDIDATE_DELAYED until reqtime is reached.
+ *   - Announcemnets that have reached their reqtime but not been requested will be either CANDIDATE_READY or
+ *     CANDIDATE_BEST
+ *   - When requested, an announcement will be in state REQUESTED until expiry is reached.
+ *   - If expiry is reached, or the peer replies to the request (either with NOTFOUND or the tx), the state becomes
+ *     COMPLETED
+ */
+enum class State : uint8_t {
+    /** A CANDIDATE announcement whose reqtime is in the future. */
+    CANDIDATE_DELAYED,
+    /** The best CANDIDATE for a given txhash; only if there is no REQUESTED announcement already for that txhash.
+     *  The CANDIDATE_BEST is the lowest-priority announcement among all CANDIDATE_READY (and _BEST) ones for that
+     *  txhash. */
+    CANDIDATE_BEST,
+    /** A REQUESTED announcement. */
+    REQUESTED,
+    /** A CANDIDATE announcement that's not CANDIDATE_DELAYED or CANDIDATE_BEST. */
+    CANDIDATE_READY,
+    /** A COMPLETED announcement. */
+    COMPLETED,
```

Comment:
That would be a more logical ordering. Let me try what would be needed to make this work later.

Reply:
This was indeed not a problem at all. I've applied the following diff:

```patch
diff --git a/src/test/fuzz/txrequest.cpp b/src/test/fuzz/txrequest.cpp
index 0ff00d23e0..ff32de25eb 100644
--- a/src/test/fuzz/txrequest.cpp
+++ b/src/test/fuzz/txrequest.cpp
@@ -129,7 +129,7 @@ class Tester
             if (ann.m_state == State::REQUESTED) return -1;
             // If it's a viable candidate, see if it has lower priority than the best one so far.
             if (ann.m_state == State::CANDIDATE && ann.m_time <= m_now) {
-                if (ret == -1 || ann.m_priority < ret_priority) {
+                if (ret == -1 || ann.m_priority > ret_priority) {
                     std::tie(ret, ret_priority) = std::tie(peer, ann.m_priority);
                 }
             }
@@ -252,7 +252,7 @@ public:
             }
             // And delete txids with only COMPLETED announcements left.
             Cleanup(txhash);
-            // CANDIDATEs for which this announcement has the lowest priority get returned.
+            // CANDIDATEs for which this announcement has the highest priority get returned.
             const Announcement& ann = m_announcements[txhash][peer];
             if (ann.m_state == State::CANDIDATE && GetSelected(txhash) == peer) {
                 result.emplace_back(ann.m_sequence, txhash, ann.m_is_wtxid);
diff --git a/src/test/txrequest_tests.cpp b/src/test/txrequest_tests.cpp
index a84dfe67f9..39a4f86ea8 100644
--- a/src/test/txrequest_tests.cpp
+++ b/src/test/txrequest_tests.cpp
@@ -162,8 +162,8 @@ public:
     /** Generate a random txhash, whose priorities for certain peers are constrained.
      *
      * For example, NewTxHash({{p1,p2,p3},{p2,p4,p5}}) will generate a txhash T such that both:
-     *  - priority(p1,T) < priority(p2,T) < priority(p3,T)
-     *  - priority(p2,T) < priority(p4,T) < priority(p5,T)
+     *  - priority(p1,T) > priority(p2,T) > priority(p3,T)
+     *  - priority(p2,T) > priority(p4,T) > priority(p5,T)
      * where priority is the predicted internal TxRequestTracker's priority, assuming all announcements
      * are within the same preferredness class.
      */
@@ -178,7 +178,7 @@ public:
                 for (size_t pos = 1; pos < order.size(); ++pos) {
                     uint64_t prio_prev = m_runner.txrequest.ComputePriority(ret, order[pos - 1], true);
                     uint64_t prio_cur = m_runner.txrequest.ComputePriority(ret, order[pos], true);
-                    if (prio_prev >= prio_cur) {
+                    if (prio_prev <= prio_cur) {
                         ok = false;
                         break;
                     }
diff --git a/src/txrequest.cpp b/src/txrequest.cpp
index d07a5203cd..a5455e2305 100644
--- a/src/txrequest.cpp
+++ b/src/txrequest.cpp
@@ -38,14 +38,14 @@ namespace {
 enum class State : uint8_t {
     /** A CANDIDATE announcement whose reqtime is in the future. */
     CANDIDATE_DELAYED,
+    /** A CANDIDATE announcement that's not CANDIDATE_DELAYED or CANDIDATE_BEST. */
+    CANDIDATE_READY,
     /** The best CANDIDATE for a given txhash; only if there is no REQUESTED announcement already for that txhash.
-     *  The CANDIDATE_BEST is the lowest-priority announcement among all CANDIDATE_READY (and _BEST) ones for that
+     *  The CANDIDATE_BEST is the highest-priority announcement among all CANDIDATE_READY (and _BEST) ones for that
      *  txhash. */
     CANDIDATE_BEST,
     /** A REQUESTED announcement. */
     REQUESTED,
-    /** A CANDIDATE announcement that's not CANDIDATE_DELAYED or CANDIDATE_BEST. */
-    CANDIDATE_READY,
     /** A COMPLETED announcement. */
     COMPLETED,
 };
@@ -106,7 +106,7 @@ using Priority = uint64_t;
 
 /** A functor with embedded salt that computes priority of an announcement.
  *
- * Lower priorities are selected first.
+ * Higher priorities are selected first.
  */
 class PriorityComputer {
     const uint64_t m_k0, m_k1;
@@ -118,7 +118,7 @@ public:
     Priority operator()(const uint256& txhash, NodeId peer, bool preferred) const
     {
         uint64_t low_bits = CSipHasher(m_k0, m_k1).Write(txhash.begin(), txhash.size()).Write(peer).Finalize() >> 1;
-        return low_bits | uint64_t{!preferred} << 63;
+        return low_bits | uint64_t{preferred} << 63;
     }
 
     Priority operator()(const Announcement& ann) const
@@ -244,10 +244,10 @@ struct TxHashInfo
     size_t m_candidate_best = 0;
     //! Number of REQUESTED announcements for this txhash.
     size_t m_requested = 0;
-    //! The priority of the CANDIDATE_BEST announcement if one exists, or min() otherwise.
-    Priority m_priority_candidate_best = std::numeric_limits<Priority>::min();
-    //! The lowest priority of all CANDIDATE_READY announcements (or max() if none exist).
-    Priority m_priority_best_candidate_ready = std::numeric_limits<Priority>::max();
+    //! The priority of the CANDIDATE_BEST announcement if one exists, or max() otherwise.
+    Priority m_priority_candidate_best = std::numeric_limits<Priority>::max();
+    //! The lowest priority of all CANDIDATE_READY announcements (or min() if none exist).
+    Priority m_priority_best_candidate_ready = std::numeric_limits<Priority>::min();
     //! All peers we have an announcement for this txhash for.
     std::vector<NodeId> m_peers;
 };
@@ -288,7 +288,7 @@ std::map<uint256, TxHashInfo> ComputeTxHashInfo(const Index& index, const Priori
             info.m_priority_candidate_best = computer(ann);
         }
         if (ann.GetState() == State::CANDIDATE_READY) {
-            info.m_priority_best_candidate_ready = std::min(info.m_priority_best_candidate_ready, computer(ann));
+            info.m_priority_best_candidate_ready = std::max(info.m_priority_best_candidate_ready, computer(ann));
         }
         // Also keep track of which peers this txhash has an announcement for (so we can detect duplicates).
         info.m_peers.push_back(ann.m_peer);
@@ -341,7 +341,7 @@ public:
             // If there is both a CANDIDATE_READY and a CANDIDATE_BEST announcement, the CANDIDATE_BEST one must be
             // at least as good (equal or lower priority) as the best CANDIDATE_READY.
             if (info.m_candidate_ready && info.m_candidate_best) {
-                assert(info.m_priority_candidate_best <= info.m_priority_best_candidate_ready);
+                assert(info.m_priority_candidate_best >= info.m_priority_best_candidate_ready);
             }
 
             // No txhash can have been announced by the same peer twice.
@@ -399,21 +399,21 @@ private:
         // Convert CANDIDATE_DELAYED to CANDIDATE_READY first.
         Modify<ByTxHash>(it, [](Announcement& ann){ ann.SetState(State::CANDIDATE_READY); });
         // The following code relies on the fact that the ByTxHash is sorted by txhash, and then by state (first
-        // _DELAYED, then _BEST/REQUESTED, then _READY). Within the _READY announcements, the best one (lowest
-        // priority) comes first. Thus, if an existing _BEST exists for the same txhash that this announcement may
-        // be preferred over, it must immediately precede the newly created _READY.
-        if (it == m_index.get<ByTxHash>().begin() || std::prev(it)->m_txhash != it->m_txhash ||
-            std::prev(it)->GetState() == State::CANDIDATE_DELAYED) {
+        // _DELAYED, then _READY, then _BEST/REQUESTED). Within the _READY announcements, the best one (highest
+        // priority) comes last. Thus, if an existing _BEST exists for the same txhash that this announcement may
+        // be preferred over, it must immediately follow the newly created _READY.
+        auto it_next = std::next(it);
+        if (it_next == m_index.get<ByTxHash>().end() || it_next->m_txhash != it->m_txhash ||
+            it_next->GetState() == State::COMPLETED) {
             // This is the new best CANDIDATE_READY, and there is no IsSelected() announcement for this txhash
             // already.
             Modify<ByTxHash>(it, [](Announcement& ann){ ann.SetState(State::CANDIDATE_BEST); });
-        } else if (std::prev(it)->GetState() == State::CANDIDATE_BEST) {
-            Priority priority_old = m_computer(*std::prev(it));
+        } else if (it_next->GetState() == State::CANDIDATE_BEST) {
+            Priority priority_old = m_computer(*it_next);
             Priority priority_new = m_computer(*it);
-            if (priority_new < priority_old) {
+            if (priority_new > priority_old) {
                 // There is a CANDIDATE_BEST announcement already, but this one is better.
-                auto new_ready_it = std::prev(it);
-                Modify<ByTxHash>(new_ready_it, [](Announcement& ann){ ann.SetState(State::CANDIDATE_READY); });
+                Modify<ByTxHash>(it_next, [](Announcement& ann){ ann.SetState(State::CANDIDATE_READY); });
                 Modify<ByTxHash>(it, [](Announcement& ann){ ann.SetState(State::CANDIDATE_BEST); });
             }
         }
@@ -425,14 +425,13 @@ private:
     {
         assert(new_state == State::COMPLETED || new_state == State::CANDIDATE_DELAYED);
         assert(it != m_index.get<ByTxHash>().end());
-        if (it->IsSelected()) {
-            auto it_next = std::next(it);
-            // The next best CANDIDATE_READY, if any, immediately follows the REQUESTED or CANDIDATE_BEST
+        if (it->IsSelected() && it != m_index.get<ByTxHash>().begin()) {
+            auto it_prev = std::prev(it);
+            // The next best CANDIDATE_READY, if any, immediately preceeds the REQUESTED or CANDIDATE_BEST
             // announcement in the ByTxHash index.
-            if (it_next != m_index.get<ByTxHash>().end() && it_next->m_txhash == it->m_txhash &&
-                it_next->GetState() == State::CANDIDATE_READY) {
+            if (it_prev->m_txhash == it->m_txhash && it_prev->GetState() == State::CANDIDATE_READY) {
                 // If one such CANDIDATE_READY exists (for this txhash), convert it to CANDIDATE_BEST.
-                Modify<ByTxHash>(it_next, [](Announcement& ann){ ann.SetState(State::CANDIDATE_BEST); });
+                Modify<ByTxHash>(it_prev, [](Announcement& ann){ ann.SetState(State::CANDIDATE_BEST); });
             }
         }
         Modify<ByTxHash>(it, [new_state](Announcement& ann){ ann.SetState(new_state); });
```